xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-filters/xwiki-platform-notifications-filters-api/src/main/java/org/xwiki/notifications/filters/internal/DefaultNotificationFilterManager.java

```diff
@@ -111,21 +113,80 @@
         Map<String, Boolean> filterActivations = getToggeableFilterActivations(user);
+        boolean prefilteringEnabled = this.configuration.isEventPrefilteringEnabled();
         Iterator<NotificationFilter> it = filters.iterator();
         while (it.hasNext()) {
             NotificationFilter filter = it.next();
-            Boolean filterActivation = filterActivations.get(filter.getName());
-            if (onlyEnabled && filterActivation != null && !filterActivation.booleanValue()) {
+            boolean filterActivation = filterActivations.getOrDefault(filter.getName(), true);
+            if (!this.shouldFilterBeSelected(filter, filterActivation, onlyEnabled, filteringPhase,
+                prefilteringEnabled)) {
                 it.remove();
             }
         }
         return filters;
     }
 
+    private boolean shouldFilterBeSelected(NotificationFilter filter, boolean filterActivation, boolean onlyEnabled,
+        NotificationFilter.FilteringPhase filteringPhase, boolean onlyExactPhase)
+    {
+        return shouldFilterBeSelectedBasedOnStatus(filterActivation, onlyEnabled)
+            && shouldFilterBeSelectedBasedOnFilteringPhase(filter, filteringPhase, onlyExactPhase);
+    }
+
+    private boolean shouldFilterBeSelectedBasedOnStatus(boolean filterActivation, boolean onlyEnabled)
+    {
+        boolean result = false;
+        // if we don't care to get only the enabled one, we can return all of them.
+        if (!onlyEnabled) {
+            result = true;
+        // else we just need to ensure that the filter is activated
+        } else if (filterActivation) {
+            result = true;
+        }
+        return result;
+    }
+
+    private boolean shouldFilterBeSelectedBasedOnFilteringPhase(NotificationFilter filter,
+        NotificationFilter.FilteringPhase filteringPhase, boolean prefilteringEnabled)
+    {
+        boolean result;
+        Set<NotificationFilter.FilteringPhase> filterFilteringPhases = filter.getFilteringPhases();
+        // by definition if the filteringPhase provided is null we return all filters
+        if (filteringPhase == null) {
+            result = true;
+        // just bulletproof in case a filter didn't properly override the method
+        } else if (filterFilteringPhases == null) {
+            result = false;
+        // if pre-filtering is enabled and we're requesting for post-filtering phases, we return only
+        // the filters that only supports post-filtering: we don't need the other ones since they have been already
+        // applied.
+        } else if (prefilteringEnabled && filteringPhase == NotificationFilter.FilteringPhase.POST_FILTERING) {
+            result = Collections.singleton(NotificationFilter.FilteringPhase.POST_FILTERING)
+                .equals(filterFilteringPhases);
+        // if pre-filtering is not enabled, but we request pre-filtering filters only, it doesn't really make sense
```

Comment:
I think I would simply apply the standard logic and return filters which only support prefiltering. Of course in practice there is none right now (but there could be soon) and more importantly no code is supposed to ask for it in postfiltering mode so it does not matter that much.

Reply:
> more importantly no code is supposed to ask for it in postfiltering mode

Yeah actually it sounded to me that it would have been almost a bug to perform such request hence this decision to not return anything for now.